File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
     int ans ;
 int cnt;
 void traverse(TreeNode* node, int B){
     if(node==NULL || ans!=-1)return;
     // if(node->left==NULL && node->right==NULL){
     //     cnt++;
     //     cout<<cnt<<", node val ="<<node->val<<endl;
     //     if(cnt==B){
     //         ans = node->val;
     //         cout<<"assigning ans to "<<node->val<<endl;
     //         return;
     //     }
     // }
     if(node->left!=NULL)
        traverse(node->left,B);
     cnt++;
     cout<<cnt<<", node val ="<<node->val<<endl;
     if(cnt==B){
        ans = node->val;
        return;
     }
     if(node->right!=NULL)
        traverse(node->right,B);
     
     
 }
    int kthSmallest(TreeNode* root, int k) {
        ans=-1;
    cnt=0;
    traverse(root,k);
    return ans;
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans=-1;
    int curr = 0;
    int k=0;
    void traverse(TreeNode root){
        if(root==null)return;
        traverse(root.left);
        // System.out.println("val ="+root.val);
        curr++;
        if(curr==k){
            ans = root.val;
            return;
        }
        traverse(root.right);
          
    }
    public int kthSmallest(TreeNode root, int kv) {
        k=kv;
        traverse(root);
        return ans;
        
    }
}